prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"compress/gzip"
+	"context"
+	"io/ioutil"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// GCSArtifact represents some output of a prow job stored in GCS
+type GCSArtifact struct {
+	// The handle of the object in GCS
+	handle *storage.ObjectHandle
+
+	// The direct link to the Artifact, can be used for read operations
+	link string
+
+	// The path of the Artifact within the job
+	path string
+
+	// Max size to read before failing
+	sizeLimit int64
+}
+
+// NewGCSArtifact returns a new GCSArtifact with a given handle, canonical link, and path within the job
+func NewGCSArtifact(handle *storage.ObjectHandle, link string, path string) *GCSArtifact {
+	return &GCSArtifact{
+		handle:    handle,
+		link:      link,
+		path:      path,
+		sizeLimit: 500e6, // 500MB max file size
+	}
+}
+
+// Size returns the size of the artifact in GCS
+func (a *GCSArtifact) Size() int64 {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
+		logrus.WithError(err).Error("Could not retrieve object attributes", err)
+	}
+	return attrs.Size
+}
+
+// JobPath gets the GCS path of the artifact within the current job
+func (a *GCSArtifact) JobPath() string {
+	return a.path
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
+	return a.link
+}
+
+// Read reads len(p) bytes from a file in GCS
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	if a.gzipped() {
+		return 0, viewers.ErrUnsupportedOp
+	}
+	reader, err := a.handle.NewRangeReader(context.Background(), off, int64(len(p)))
+	if err != nil {
+		return 0, err
+	}
+	n, err = reader.Read(p)
+	reader.Close()
+	return n, err
+}
+
+// ReadAtMost reads at most n bytes from a file in GCS. If the file is compressed (gzip) in GCS, n bytes
+// of gzipped content will be downloaded and decompressed into potentially GREATER than n bytes of content.
+func (a *GCSArtifact) ReadAtMost(n int64) ([]byte, error) {
+	reader, err := a.handle.NewRangeReader(context.Background(), 0, n+1)
+	if err != nil {
+		return []byte{}, err
+	}
+	var p []byte
+	var e error
+	if a.gzipped() {
+		gReader, err := gzip.NewReader(reader)
```

Comment:
comment about gzip composition outside of gcs outstanding

Reply:
transcoding to gzip of blobs is a GCS detail. this isn't about handling all gzipping. when they're transparently gzipped you can only read from the beginning and the artifact consumer can reasonably expect them to be un-gzipped. for an artifact that was actually gzipped prior to upload we should not magically ungzip it.

https://cloud.google.com/storage/docs/transcoding

Reply:
SGTM